### PR TITLE
No interfaces

### PR DESCRIPTION
**Pull request by buybackoff**
These changes combined push IPC performance to 22 Mops from 14 Mops. It requires some cleaning, so this PR is mostly FYI & WIP. #29 #31

### CODE REVIEW

**Comment by buybackoff** [comment]
With the https://github.com/AdaptiveConsulting/Aeron.NET/pull/37/commits/0f9c0b32fb2e040aef87451718cd338aa1c1caeb commit, the median values for throughput printed lines:
* Non-cached delegate: 22.6 Mops, StDev 0.63, a lot of GC0 for temp delegates
* Cached delegate: 19.9 Mops, StDev 0.18, no garbage at all
* Current master: 13.7 Mops, StDev 0.31, a lot of GC0 for temp delegates

On the same machine, Java version runs at median 23.7 Mops, StDev 0.90.

Main contributions to performance are:
* Interface method calls elimination - other than they are more expensive by themselves, they are not candidates for JIT inlining. This change alone in the main contributor, it boosts throughput to c.20 Mops.
* MemoryCopy was implemented as a DynamicMethod delegate, which is also not a candidate for JIT inlining. Manual memory copy inside `UnsafeBuffer.PutBytes` gives noticeable improvement. (This [post](http://frankniemeyer.blogspot.ru/2014/07/methods-for-reading-structured-binary.html) discusses different memcpy methods, simple "*dst++ = *src++" is already one of the fastest. In the latest commit memcpy was optimized to copy in chunks of 64/32/16/8/4/1 bytes).
* Making UnsafeBuffer sealed and implementing header flyweights as structs, sealing virtual methods in other places.



There is a curios case with a cached delegate that reduces allocations but performance drops with that, I have posted a [question](http://stackoverflow.com/questions/37372719/c-sharp-why-using-instance-method-as-delegate-allocates-gc0-temp-objects-but-10) on SO about this issue.

Ping-pong latency is not affected by delegate caching (that affects only IpcThroughput project):

Pinging 1000000 messages
Histogram of RTT latencies in microseconds.
       Value     Percentile TotalCount 1/(1-Percentile)

      10.591 0.000000000000         50           1.00
      11.871 0.100000000000     301357           1.11
      11.871 0.200000000000     301357           1.25
      11.871 0.300000000000     301357           1.43
      12.191 0.400000000000     437437           1.67
      12.831 0.500000000000     678836           2.00
      12.831 0.550000000000     678836           2.22
      12.831 0.600000000000     678836           2.50
      12.831 0.650000000000     678836           2.86
      13.151 0.700000000000     930516           3.33
      13.151 0.750000000000     930516           4.00
      13.151 0.775000000000     930516           4.44
      13.151 0.800000000000     930516           5.00
      13.151 0.825000000000     930516           5.71
      13.151 0.850000000000     930516           6.67
      13.151 0.875000000000     930516           8.00
      13.151 0.887500000000     930516           8.89
      13.151 0.900000000000     930516          10.00
      13.151 0.912500000000     930516          11.43
      13.151 0.925000000000     930516          13.33
      13.471 0.937500000000     953262          16.00
      13.471 0.943750000000     953262          17.78
      13.471 0.950000000000     953262          20.00
      13.791 0.956250000000     961820          22.86
      14.119 0.962500000000     972590          26.67
      14.119 0.968750000000     972590          32.00
      14.119 0.971875000000     972590          35.56
      14.439 0.975000000000     976122          40.00
      15.079 0.978125000000     978494          45.71
      16.687 0.981250000000     981566          53.33
      20.863 0.984375000000     984523          64.00
      24.703 0.985937500000     985997          71.11
      29.839 0.987500000000     987563          80.00
      35.935 0.989062500000     989113          91.43
      44.287 0.990625000000     990654         106.67
      61.279 0.992187500000     992190         128.00
      75.071 0.992968750000     992978         142.22
      95.935 0.993750000000     993751         160.00
     117.759 0.994531250000     994614         182.86
     120.639 0.995312500000     995320         213.33
     125.759 0.996093750000     996106         256.00
     129.279 0.996484375000     996510         284.44
     134.399 0.996875000000     996896         320.00
     142.207 0.997265625000     997276         365.71
     153.343 0.997656250000     997664         426.67
     169.727 0.998046875000     998049         512.00
     182.527 0.998242187500     998244         568.89
     201.215 0.998437500000     998439         640.00
     221.695 0.998632812500     998634         731.43
     258.943 0.998828125000     998830         853.33
     324.095 0.999023437500     999024        1024.00
     364.287 0.999121093750     999122        1137.78
     396.287 0.999218750000     999220        1280.00
     422.911 0.999316406250     999317        1462.86
     446.207 0.999414062500     999416        1706.67
     468.735 0.999511718750     999512        2048.00
     479.231 0.999560546875     999562        2275.56
     490.239 0.999609375000     999612        2560.00
     503.039 0.999658203125     999659        2925.71
     525.311 0.999707031250     999708        3413.33
     544.767 0.999755859375     999761        4096.00
     552.447 0.999780273438     999782        4551.11
     564.223 0.999804687500     999805        5120.00
     587.775 0.999829101563     999831        5851.43
     606.207 0.999853515625     999854        6826.67
     641.535 0.999877929688     999878        8192.00
     660.479 0.999890136719     999891        9102.22
     694.271 0.999902343750     999903       10240.00
     741.887 0.999914550781     999915       11702.86
     786.943 0.999926757813     999927       13653.33
     889.855 0.999938964844     999939       16384.00
     963.071 0.999945068359     999946       18204.44
    1018.367 0.999951171875     999952       20480.00
    1104.895 0.999957275391     999958       23405.71
    1205.247 0.999963378906     999964       27306.67
    1394.687 0.999969482422     999970       32768.00
    1618.943 0.999972534180     999973       36408.89
    1657.855 0.999975585938     999976       40960.00
    1854.463 0.999978637695     999979       46811.43
    1946.623 0.999981689453     999982       54613.33
    1993.727 0.999984741211     999985       65536.00
    2004.991 0.999986267090     999987       72817.78
    2055.167 0.999987792969     999988       81920.00
    2081.791 0.999989318848     999990       93622.86
    2084.863 0.999990844727     999991      109226.67
    2093.055 0.999992370605     999993      131072.00
    2119.679 0.999993133545     999994      145635.56
    2119.679 0.999993896484     999994      163840.00
    2207.743 0.999994659424     999995      187245.71
    2226.175 0.999995422363     999996      218453.33
    2861.055 0.999996185303     999997      262144.00
    2861.055 0.999996566772     999997      291271.11
    2861.055 0.999996948242     999997      327680.00
    2889.727 0.999997329712     999998      374491.43
    2889.727 0.999997711182     999998      436906.67
    3211.263 0.999998092651     999999      524288.00
    3211.263 0.999998283386     999999      582542.22
    3211.263 0.999998474121     999999      655360.00
    3211.263 0.999998664856     999999      748982.86
    3211.263 0.999998855591     999999      873813.33
    3887.103 0.999999046326    1000000
    3887.103 1.000000000000    1000000
#[Mean    =       14.064, StdDeviation   =       22.107]
#[Max     =     3887.103, Total count    =      1000000]
#[Buckets =           24, SubBuckets     =         2048]

**Comment by buybackoff** [comment]
It was a fun challenge, but at the same time quite tiresome to do JIT compiler job for it... Obviously .NET JIT is lazy with many things. I will now stop adding any new changes, current version is on par with Java. This is a kind of kitchen-sink commit, so please cherry pick the parts you think are appropriate, or suggest what other changes are needed to merge it?

**Comment by JPWatson** [comment]
Nice work 👍 , may take a bit of time to go through this :)

**Comment by buybackoff** [comment]
Final touches. PerfView showed which methods were still not inlined by JIT. Now the stats looks near pristine:
Only those methods are called:
image

And only FragmentHandler is allocated:
image

With cached FragmentHandler performance is c.28 Mops, all methods are still inlined (Subsciber.Run becomes inlined instead of TermReader.Read) and no allocations at all. Yesterdays there was a talk on YouTube about GC and the fact that it allocates more than requested - in 8Kb chunks (or something). Probably allocation somehow improves cache locality.

**Comment by JPWatson** [comment]
Hi Victor, 

Really appreciate the effort you've put in, sorry it's taken a while for me to reply.

At the moment, when I reinstate some of the virtual methods, tests are failing.
I think it's important that we find a way to keep the tests running going forward. I see a 3 options:
a. Find a open-source/free version of a mocking framework that allows mocking at the IL level, removing the need of virtual methods & interfaces like typemock.
b. #if preprocessor directives all over the place.
c. Keep virtual methods & interfaces.
I haven't found any open-source versions of a., there's such a performance boost from not having the virtual/interface call & the JIT optimisations available as a result that c. isn't a very good option either. b. is going to be a mess but I'm going to give it a go.

Second point is, we want to keep divergence from the Java code to a minimum, otherwise maintenance is going to end up being impossible. For instance the changes in TermAppender are significant enough that it's potentially going to cause issues when porting any changes in futures. And I'm not sure changing DataHeaderFlyweight to a struct changes much and it impacts the the code quite a bit.  

If you've got any thoughts let me know.

**Comment by buybackoff** [comment]
Why at all something needs to be mocked in such a small project? Why not just create real temporary instances of buffers, logs, etc. Some private members could be made internal and visible to a test assembly if needed.

**Comment by odeheurles** [comment]
The tests have been ported directly from the java codebase. it was quite time consuming to port the tests (probably the most time consuming part actually) and changing the testing approach would take significantly more time - the original codebase use mocks quite a lot, especially for UnsafeBuffer. Removing virtual method calls and interface breaks 100+ unit tests, about half of the tests of the whole project, this is what I tried to point out in my [comment](https://github.com/AdaptiveConsulting/Aeron.NET/issues/31#issuecomment-220364345) earlier. If you look at the Aeron client tests you will see lots of tests checking that the relevant methods are called on UnsafeBuffer and in the correct order. 

We are at an early stage in the project and those tests give us some confidence about the correctness of the port, I do not think it would be a good idea to drop them at this stage to get a few more Mops.

I'm sure there are quite a few of your changes which could be merged incrementally, without breaking the tests or drifting too far away from the original codebase. I could not stress enough how painful it will be on the long run to maintain this port if we derive too far away from the java codebase. Martin and Todd will continue to push new commits and we do not want to endup in a place where it takes a lot of time and effort to port those changes across or risk to introduce errors in the .NET codebase because the 2 codebase have drifted too far away from each other.

We've not stated any guidelines regarding contributions, but if I had a first go at it I would say:
1. make sure it does not break correctness and tests are passing with the PR
2. try to keep the code as close as possible from the original java source, that's very important from a code maintainability point of view as we intend to maintain Aeron.NET on the long run and port new Aeron features going forward
3. any performance or other improvement is welcome as long as it does not break 1 or 2. If the improvement is significant we can consider breaking (2), but this should be discussed with maintainers first.

Hope this makes sense. There is some very good stuff in your PR and we are keen to get those changes in but I think we should have a chat together to figure out what is the best way to do that and which changes we should bring in or not. Should we maybe have a skype/hangout session to discuss? That might be easier than here :)

**Comment by buybackoff** [comment]
I realize that there is no way this PR could be merged as is and I am not pushing this. I noted in the beginning that it is WIP and more experimental just to discuss ways to improve performance. It is very likely that I will continue to use my original fork for IPC with incompatible modification that allows term writers to unblock without a conductor. For networking the performance of this project is already decent and any changes could be introduced gradually. Latency improvement is quite marginal after all these changes.

Probably the most important thing for me in the context of long-running services was to profile allocations, confirm that there were almost none of them and reduce garbage to absolute zero with this commit https://github.com/AdaptiveConsulting/Aeron.NET/pull/37/commits/75e9be24e4b6d871759a963df1f344db96b7cfe4. Also manual MemCopy gives visible gains. And the second fence in the HeaderWriter is really not needed. Everything else could wait until the original codebase is stable or longer.

Maybe base classes with virtual methods could work better than interfaces - there is work in progress to make sealed methods JIT-inlineable: https://github.com/dotnet/coreclr/issues/1166 and microbenchmarks show that they are still faster than interfaces.

File: src/Adaptive.Aeron/LogBuffer/HeaderWriter.cs [hunk 1]
```diff
@@ -35,18 +33,15 @@ public HeaderWriter(IDirectBuffer defaultHeader)
...
             var termOffsetSessionId = _sessionId | (uint)offset;
             var streamAndTermIds = _streamId | ((long)termId << 32);
 
-            //TODO why not just putlongvolatile?
-            termBuffer.PutLongOrdered(offset + HeaderFlyweight.FRAME_LENGTH_FIELD_OFFSET, lengthVersionFlagsType);
-            Thread.MemoryBarrier();
-
-            termBuffer.PutLong(offset + DataHeaderFlyweight.TERM_OFFSET_FIELD_OFFSET, termOffsetSessionId);
-            termBuffer.PutLong(offset + DataHeaderFlyweight.STREAM_ID_FIELD_OFFSET, streamAndTermIds);
+            *(int*)(termBuffer.BufferPointer + offset + HeaderFlyweight.VERSION_FIELD_OFFSET) = _versionFlagsType;
```
Comment (JPWatson) [inline src/Adaptive.Aeron/LogBuffer/HeaderWriter.cs]: @buybackoff where's the memory barrier?

Comment (buybackoff) [inline src/Adaptive.Aeron/LogBuffer/HeaderWriter.cs]: In my final version, it is moved into TermAppender with a direct `Volatile.Write` call. Original PutLongOrdered is implemented the same way via `Volatile.Write` [here](https://github.com/AdaptiveConsulting/Aeron.NET/blob/ba58ca2e85c215f85c9c65b4e3fa9a80402eafb7/src/Adaptive.Agrona/Concurrent/UnsafeBuffer.cs#L296).

**Comment by buybackoff** [comment]
There is another possibility to keep interfaces but get devirtualization of calls - constrained generics with structs. I googled for "C# devirtualization" after reading [this article](https://www.infoq.com/news/2016/06/systems-programming-qcon?utm_source=infoq&utm_medium=popular_widget&utm_campaign=popular_content_list&utm_content=homepage) and it looks like constrained calls to structs are always devirtualized because structs do not support virtual methods and the compiler could safely replace an interface call with the actual method. A simple [test](https://github.com/Spreads/Spreads/commit/82a0f7ba30f9ffa1cd4a568a3d156756f6caf354#diff-802be01990ce8ad2d6a36a3fcb106193R200) confirms that there are no overheads:

Value 2162 (1000000000)
Struct 2187
Struct as Interface 4575
**Constrained struct 2175**
Static Class 2293
Class 2173
Sealed Class 2172
Class as Interface 3335
Derived Class 2628
Derived Class as Interface 3164
Sealed Derived Class 3188
Sealed Derived Class as Interface 3389

**Comment by buybackoff** [comment]
Could you please confirm if HeaderWriter.Write needs to be virtual for tests? I understand the issue with UnsafeBuffer and tests, but cannot find any place where HeaderWriter is inherited. Performance impact is quite small, but slightly visible.

**Comment by JPWatson** [comment]
Yeah it does need to be virtual. It's mocked in TermAppenderTest.

**Comment by JPWatson** [comment]
Have you got all the meat from this PR? (Apart from the virtual methods which has its own issue)

**Comment by buybackoff** [comment]
Yes, other than interfaces et al. everything is in master already. I was going to write about this here and discuss potential further steps, but it was too late... Wanted to do tomorrow.

In brief, I want to try the constrained generic thing. It will require passing UnsafeBuffer by ref everywhere and changing signatures to constrained generics, but it is the only way to keep interfaces and avoid overheads. I am not familiar with the mocking framework you use, so not sure if that will work for tests. If it will work, we will get all the performance (unless JIT doesn't like structs) and keep interfaces. But UnsafeBuffer will become a struct, and two its inheritors will have it as a field.

We could continue our discussion here even if the issue is closed, so please close it.

**Comment by buybackoff** [comment]
A possible solution is to create another implementation as a struct, and use the existing class in tests.

**Comment by JPWatson** [comment]
I'm closing this now, see #45, we can continue the discussion there.

**Pull request closed without merging**